cmd/dev/snapshots.cpp

```diff
@@ -430,6 +436,63 @@ void open_btree_index(const SnapshotSubcommandSettings& settings) {
     SILK_INFO << "Open btree index elapsed: " << duration_as<std::chrono::milliseconds>(elapsed) << " msec";
 }
 
+void open_existence_index(const SnapshotSubcommandSettings& settings) {
+    ensure(!settings.input_file_path.empty(), "open_existence_index: --file must be specified");
+    ensure(settings.input_file_path.extension() == ".kv", "open_existence_index: --file must be .kv file");
+
+    std::filesystem::path existence_index_file_path = settings.input_file_path;
+    existence_index_file_path.replace_extension(".kvei");
+    SILK_INFO << "KV file: " << settings.input_file_path.string() << " KVEI file: " << existence_index_file_path.string();
+    const auto salt_path = existence_index_file_path.parent_path().parent_path() / "salt-state.txt";
+    std::ifstream salt_stream{salt_path, std::ios::in | std::ios::binary};
+    std::array<char, sizeof(uint32_t)> salt_bytes{};
+    salt_stream.read(salt_bytes.data(), salt_bytes.size());
+    const uint32_t salt = endian::load_big_u32(reinterpret_cast<uint8_t*>(salt_bytes.data()));
+    SILK_INFO << "Snapshot salt: " << salt;
+    std::chrono::time_point start{std::chrono::steady_clock::now()};
+    seg::Decompressor kv_decompressor{settings.input_file_path};
+    kv_decompressor.open();
+    snapshots::index::ExistenceIndex::Reader existence_index{existence_index_file_path};
+    SILK_INFO << "Starting KV scan and existence index check";
+
+    struct Hash128 {      // TODO(canepat) move recsplit definition and use that
+        uint64_t first;   // The high 64-bit hash half
+        uint64_t second;  // The low 64-bit hash half
+
+        bool operator<(const Hash128& o) const { return first < o.first || second < o.second; }
+    };
+
+    size_t matching_count{0}, key_count{0};
+    bool is_key{true};
+    Bytes key, value;
+    auto kv_iterator = kv_decompressor.begin();
+    while (kv_iterator != kv_decompressor.end()) {
+        if (is_key) {
+            key = *kv_iterator;
+            ++key_count;
+        } else {
+            value = *kv_iterator;
+            Hash128 key_hash;
+            snapshots::rec_split::MurmurHash3_x64_128(key.data(), key.size(), salt, &key_hash);
+            const bool key_found = existence_index.contains_hash(key_hash.first);
+            SILK_DEBUG << "KV: key=" << to_hex(key) << " value=" << to_hex(value);
+            ensure(key_found,
```

Comment:
It does check existence as well, but only probabilistically. An alternative name could be: `contains_hash_perhaps`, or if we invert it `not_contains_hash_surely`, but neither of 2 names captures the asymmetric nature of the result. If it was returning an enum, it could have 2 values: NOT_PRESENT and POTENTIALLY_PRESENT. I'm not sure that such change worth it compared to the current bool-based approach.

Reply:
> It's just an observation but I believe that checking existence of a key in a bloom filter is a bit confusing because it's not about presence but absence.

improved in commit https://github.com/erigontech/silkworm/pull/2386/commits/846352379268c88ce0ee4fb6482629cb536144a9